### ISSUE TITLE
Invert “if” statement to reduce nesting

### ISSUE THREAD

**Issue by marco-carvalho** [eslint/eslint#10120]
**Please describe what the rule should do:**
Invert “if” statement to reduce nesting, just like what Resharper does

**What category of rule is this? (place an "X" next to just one item)**

[X] Enforces code style
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

**BAD**
```js
function getPayAmount() {
    var result;
    if (_isDead) result = deadAmount();
    else {
        if (_isSeparated) result = separatedAmount();
        else {
            if (_isRetired) result = retiredAmount();
            else result = normalPayAmount();
        };
    }
     return result;
};
```

**GOOD**
```js
function getPayAmount() {
    if (_isDead)      return deadAmount();
    if (_isSeparated) return separatedAmount();
    if (_isRetired)   return retiredAmount();

    return normalPayAmount();
}; 
```

**Why should this rule be included in ESLint (instead of a plugin)?**
The GOOD code is much more readable than the BAD code.

**Comment by platinumazure** [comment]
This could be very tricky to detect, and I'm not sure this is critical enough to add to core. I'm mildly opposed at this point.

**Comment by Slayer95** [comment]
As a first step, it might be possible to detect the last instruction of each code path, so to  have this

```js
function getPayAmount() {
    var result;
    if (_isDead) result = deadAmount();
    else {
        if (_isSeparated) result = separatedAmount();
        else {
            if (_isRetired) result = retiredAmount();
            else result = normalPayAmount();
        };
    }
     return result;
};
```
rewrite to

```js
function getPayAmount() {
    var result;
    if (_isDead) return deadAmount();
    else {
        if (_isSeparated) return separatedAmount();
        else {
            if (_isRetired) return retiredAmount();
            else return normalPayAmount();
        };
    }
     return result;
};

```
The rest of the job can be handled by the already existing rules [no-else-return](https://eslint.org/docs/rules/no-else-return) and [no-unused-vars](https://eslint.org/docs/rules/no-unused-vars)

**Comment by platinumazure** [comment]
Hmm. I could *maybe* get behind a rule proposal where the rule could be configured for either of these situations:

1. All code paths before the return statement assign to the same variable, and that variable is immediately returned: Prefer return statements (what you want)
1. All code paths end in a return statement: Prefer assigning to a variable and having one return statement at the end (the inverse)

But that might not be super beneficial for your use case.

**Comment by platinumazure** [comment]
Closing as there does not appear to be [consensus](https://eslint.org/docs/maintainer-guide/issues#consensus) on this issue.

You are more than welcome to make a custom rule that fits your needs in the meanwhile. Please stop by our [Gitter chat](https://gitter.im/eslint/eslint) if you need help with that effort. Thanks!
